### ISSUE TITLE
Add support for Metal Backend to the RHI

### ISSUE THREAD

**Issue by Bentheminernz** [rt64/rt64#6]
Is macOS support planned for this project?

**Comment by DarioSamo** [comment]
Metal as a backend is something that could be easily implemented into the current RHI. It is a great API for this project and it'll be a natural fit.

However, I do not own Mac hardware or know enough about the platform to develop it. But someone's already expressed interest in giving it a shot.

**Comment by HopeSuffers** [comment]
@DarioSamo @dcvz I saw that you created the [feature/metal](https://github.com/rt64/rt64/activity?ref=feature/metal) branch recently. I worked with vulkan raytracing before and am currently developing on a M3 pro MacBook. If you are open I'm willing to contribute to rt64

**Comment by dcvz** [comment]
> @DarioSamo @dcvz I saw that you created the [feature/metal](https://github.com/rt64/rt64/activity?ref=feature/metal) branch recently. I worked with vulkan raytracing before and am currently developing on a M3 pro MacBook. If you are open I'm willing to contribute to rt64

Yeah, I've started the project now, I'll push some more later today. I'm first starting out without the ray tracing part but will eventually get there too. Definitely open to more contributors on the front, we don't have a server for dev yet though, I'll talk to Dario about it. In the meantime you can find me on Discord with the same username as GitHub.

**Comment by HopeSuffers** [comment]
> > @DarioSamo @dcvz I saw that you created the [feature/metal](https://github.com/rt64/rt64/activity?ref=feature/metal) branch recently. I worked with vulkan raytracing before and am currently developing on a M3 pro MacBook. If you are open I'm willing to contribute to rt64
> 
> Yeah, I've started the project now, I'll push some more later today. I'm first starting out without the ray tracing part but will eventually get there too. Definitely open to more contributors on this, we don't have a server for dev yet though, I'll talk to Dario about it. In the meantime you can find me on Discord with the same username as GitHub.

Nice. I just added you on discord :)

**Comment by IsaacMarovitz** [comment]
Nice that you guys are doing it I had started my own attempt https://github.com/rt64/rt64/compare/main...IsaacMarovitz:rt64:main

**Comment by ComplexPlane** [comment]
Maybe the ship has sailed at this point with the current Vulkan and D3D12 support, but have you ever considered targeting WebGPU? I've heard it's a great API and despite the name, allows targeting D3D12/Metal/Vulkan natively with minimal overhead. Chromium's Dawn is one implementation.

**Comment by DarioSamo** [comment]
> Maybe the ship has sailed at this point with the current Vulkan and D3D12 support, but have you ever considered targeting WebGPU?

Yes, it lacks a lot of features that RT64 requires.
- There's no hardware RT support.
- [Bindless resources aren't supported](https://github.com/gpuweb/gpuweb/issues/380).
- [No support for SPIR-V and reliance on WGSL](https://github.com/gpuweb/gpuweb/issues/847).
- [Push constants aren't supported](https://github.com/gpuweb/gpuweb/issues/75).
- [Dual Source Blending is currently in progress but not part of the specification yet, which is absolutely necessary](https://github.com/gpuweb/gpuweb/pull/4621).
- [It seems vague whether specialization constants are supported](https://github.com/gpuweb/gpuweb/issues/572).

**Comment by iMonZ** [comment]
So exited for this

**Comment by Bryceshaw06** [comment]
Any progress or updates on this?

**Comment by DarioSamo** [comment]
Completed by #9. Congratulations to everyone involved!
